GVRf/Framework/src/org/gearvrf/GVRViewManager.java

```diff
@@ -246,27 +260,144 @@ void beforeDrawEyes() {
         mFrameHandler.beforeDrawEyes();
     }
 
+    @Override
+    public void captureScreenCenter(GVRScreenshotCallback callback) {
+        mScreenshotCenterCallback = callback;
+    }
+
+    @Override
+    public void captureScreenLeft(GVRScreenshotCallback callback) {
+        mScreenshotLeftCallback = callback;
+    }
+
+    @Override
+    public void captureScreenRight(GVRScreenshotCallback callback) {
+        mScreenshotRightCallback = callback;
+    }
+
+    Bitmap generateBitmap(final byte[] byteArray, final int width,
+            final int height) {
+        int[] pixels = new int[width * height];
+        for (int row = 0; row < height; row++) {
+            int start_position = row * width;
+            int reverse_start_position = (height - 1 - row) * width;
+            for (int col = 0; col < width; col++) {
+                int position = (start_position + col) * 4;
+                int r = byteArray[position++] & 0xff;
+                int g = byteArray[position++] & 0xff;
+                int b = byteArray[position] & 0xff;
+                // flip the image vertically
+                pixels[reverse_start_position + col] = Color.rgb(r, g, b);
+            }
+        }
+        return Bitmap.createBitmap(pixels, width, height,
+                Bitmap.Config.ARGB_8888);
+    }
+
+    void returnScreenshotToCaller(final GVRScreenshotCallback callback,
+            final byte[] byteArray, final int width, final int height) {
+        // run the callback function in a background thread
+        Threads.spawn(new Runnable() {
+            public void run() {
+                final Bitmap capturedBitmap = generateBitmap(byteArray, width,
+                        height);
+                callback.onScreenCaptured(capturedBitmap);
+            }
+        });
+    }
+
     void onDrawEyeView(int eye, float fovDegrees) {
         mCurrentEye = eye;
         if (!(mSensoredScene == null || !mMainScene.equals(mSensoredScene))) {
             GVRCameraRig mainCameraRig = mMainScene.getMainCameraRig();
+
+            // if screenshot is enabled, initialize the readback buffer
+            ByteBuffer readbackBuffer = null;
+            int width = 0, height = 0;
+            if (mScreenshotCenterCallback != null
+                    || mScreenshotLeftCallback != null
+                    || mScreenshotRightCallback != null) {
+                width = mRenderBundle.getRightRenderTexture().getWidth();
+                height = mRenderBundle.getRightRenderTexture().getHeight();
```

Comment:
This is a small detail, but: that `mRenderBundle.getRightRenderTexture()` is not free, and it would be better to call it once, saving the result to a local variable.